### ISSUE TITLE
The large error for PQMF analysis and synthesis.

### ISSUE THREAD

**Issue by MaNatsu8023** [kan-bayashi/ParallelWaveGAN#195]
I load a speech signal x, 
x_ana=analysis (x)
x_out= synthesis(x)
the mse (x-x_out)**2 will become 0.0034, is it acceptable？

**Comment by kan-bayashi** [comment]
Maybe this is related to the design of prototype filter.
Current prototype filter is based on the following paper.

- `A Kaiser window approach for the design of prototype filters of cosine modulated filterbanks`: https://ieeexplore.ieee.org/abstract/document/681427

And the hyperparameters are manually decided so there is room for improvement.
Do you know a good method to decide the prototype filter?

**Comment by tosaka-m** [comment]
With regard to cutoff_ratio (which is the hyperparameter of current prototype filter), I think "A Kaiser window approach..." paper says that cutoff_ratio should be defined by the value minimizes the following objective function.
<img src=
"https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Cbegin%7Balign%2A%7D%0A%5Cphi_%7Bnew%7D+%3D+%5Ctext%7Bmax%7D_%7Bn%5Cin%5Cmathbb%7BZ%7D%2C+n%5Cneq0%7D+%7Cg%282Mn%29%7C%2C+%5C+%5Ctext%7Bwhere%7D+%5C+g+%3D+%5Ctext%7Bconvolution%7D%28h%2C+%5Chat%7Bh%7D%29%2C+%5C+%5Chat%7Bh%7D%28n%29+%3D+h%28-n%29.%0A%5Cend%7Balign%2A%7D%0A" 
alt="\begin{align*}
\phi_{new} = \text{max}_{n\in\mathbb{Z}, n\neq0} |g(2Mn)|, \ \text{where} \ g = \text{convolution}(h, \hat{h}), \ \hat{h}(n) = h(-n).
\end{align*}
">

The definition of <img src=
"https://render.githubusercontent.com/render/math?math=%5Ctextstyle+g" 
alt="g"> is based on the following guesses.
Let <img src=
"https://render.githubusercontent.com/render/math?math=%5Ctextstyle+P%28z%29%2C+%5Chat%7BP%7D%28z%29%2C+G%28z%29" 
alt="P(z), \hat{P}(z), G(z)"> be Z-transforms of <img src=
"https://render.githubusercontent.com/render/math?math=%5Ctextstyle+h%2C+%5Chat%7Bh%7D%2C+g" 
alt="h, \hat{h}, g"> (<img src=
"https://render.githubusercontent.com/render/math?math=%5Ctextstyle+P%2C+G" 
alt="P, G"> are the same as the notations in the paper). By the definitions in the paper,
<img src=
"https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Cbegin%7Balign%2A%7D%0AG%28e%5E%7Bjw%7D%29+%3D+%7CP%28e%5E%7Bjw%7D%29%7C%5E2+%3D+P%28e%5E%7Bjw%7D%29%5Coverline%7BP%28e%5E%7Bjw%7D%29%7D+%3D+P%28e%5E%7Bjw%7D%29P%28e%5E%7B-jw%7D%29+%3D+P%28e%5E%7Bjw%7D%29%5Chat%7BP%7D%28e%5E%7Bjw%7D%29.%0A%5Cend%7Balign%2A%7D%0A" 
alt="\begin{align*}
G(e^{jw}) = |P(e^{jw})|^2 = P(e^{jw})\overline{P(e^{jw})} = P(e^{jw})P(e^{-jw}) = P(e^{jw})\hat{P}(e^{jw}).
\end{align*}
"> 

So I guessed <img src=
"https://render.githubusercontent.com/render/math?math=%5Ctextstyle+G%28z%29+%3D+P%28z%29%5Chat%7BP%7D%28z%29" 
alt="G(z) = P(z)\hat{P}(z)">, and then it follows that
<img src=
"https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Cbegin%7Balign%2A%7D%0A%5Ctext%7BZ-transform%7D%28g%29+%3D+G+%3D+P%5Chat%7BP%7D+%3D+%5Ctext%7BZ-transform%7D%28h%29%5Ctext%7BZ-transform%7D%28%5Chat%7Bh%7D%29+%3D++%5Ctext%7BZ-transform%7D%28%5Ctext%7Bconvolution%7D%28h%2C+%5Chat%7Bh%7D%29%29.%0A%5Cend%7Balign%2A%7D%0A" 
alt="\begin{align*}
\text{Z-transform}(g) = G = P\hat{P} = \text{Z-transform}(h)\text{Z-transform}(\hat{h}) =  \text{Z-transform}(\text{convolution}(h, \hat{h})).
\end{align*}
">
Therefore, the definition of <img src=
"https://render.githubusercontent.com/render/math?math=%5Ctextstyle+g" 
alt="g"> should be <img src=
"https://render.githubusercontent.com/render/math?math=%5Ctextstyle+g+%3D+%5Ctext%7Bconvolution%7D%28h%2C+%5Chat%7Bh%7D%29." 
alt="g = \text{convolution}(h, \hat{h}).">

I tried to compute the value of optimized cutoff_ratio in [this colab notebook](https://colab.research.google.com/drive/16fK224dM_2hi10HMs8irebrlbOgHikE6?usp=sharing).
In sine wave case, the reconstruction with optimized cutoff_ratio is nearly perfect.

**Comment by kan-bayashi** [comment]
@tosaka-m Thank you for sharing great insight!
We should follow your optimization method.
Have you already checked the effect of the prototype filter parameters for the vocoder quality?

**Comment by tosaka-m** [comment]
No, not yet. I will try to fine tune the model sometime soon and share results.

**Comment by tosaka-m** [comment]
@kan-bayashi I have applied [these changes](https://github.com/kan-bayashi/ParallelWaveGAN/compare/master...tosaka-m:fix/pqmf) and checked synthesized audio samples in the following cases. (please download the results from [this link](https://drive.google.com/file/d/1jOM689H7vzdAHfp121HZg8Aa3xKLn9x2/view?usp=sharing).)

1. [train_nodev_jsut_config_ft.yml] = fine tuned 200k iter from jsut_multi_band_melgan.v2 1000k iter model you provided. (I used egs/jsut/voc1/run.sh to fine tune.)
2. [train_nodev_jsut_config.yml.modi] = jsut_multi_band_melgan.v2 1000k iter model (not fine tuned) and modified pqmf code.
3. [train_nodev_jsut_config.yml.prev] = jsut_multi_band_melgan.v2 1000k iter model  and orig pqmf code.

To me, the synthesized audio quality sounds the same in all cases. Could you check the results, please?

**Comment by tosaka-m** [comment]
There is Tensorboard losses for eval:
<img width="1385" alt="tensorboard_screen_shot_eval" src="https://user-images.githubusercontent.com/52565593/90311572-1105a780-df37-11ea-966a-ac995f86920c.png">

train:
<img width="1148" alt="tensorboard_screen_shot_train" src="https://user-images.githubusercontent.com/52565593/90311573-14009800-df37-11ea-8fe7-f14bb12aad48.png">


(Note that the losses at ~1M iter are results of the model you provided and  I'm not sure if train, eval data are the same as ~1M iter.)

**Comment by ahmed-fau** [comment]
@tosaka-m I have tried your edit to apply PQMF analysis-synthesis filtering to one LJ signal (LJ001-0001.wav). 

The sum of the absolute reconstruction error between original and synthesized waveform decreased from 6145.9399 (the current implementation) to 11.4774 (your implementation). That sounds awesome!

**Comment by kan-bayashi** [comment]
@tosaka-m Thank you for sharing results!
It sounds almost the same but your change is valid.
Especially, number of taps is a bug. I will fix it.

**Comment by kan-bayashi** [comment]
I checked @tosaka-m change.
<img width="673" alt="スクリーンショット 2020-08-16 午後9 06 59" src="https://user-images.githubusercontent.com/22779813/90333890-7a072100-e004-11ea-8861-fe9a3c28706b.png">
Both bug fixing and optimized parameter bringss improvement.

**Comment by tosaka-m** [comment]
@ahmed-fau @kan-bayashi Thank you for checking!
I'm glad the reconstruction error issue is mostly resolved.

**Comment by kan-bayashi** [comment]
I fixed the bug and updated the default value in #204.
Old version model can also be used.

And I added a filter parameter optimization example in notebook by following @tosaka-m example.
https://github.com/kan-bayashi/ParallelWaveGAN/blob/bc376b4f4e95d2448e19bf8ca2487ec64e8e45cc/notebooks/filter_design_example.ipynb

I will close this issue.

**Comment by hitrefresh** [comment]
@tosaka-m @kan-bayashi I had a followup question regarding the colab notebook. I calculated the MSE error (reconstructed and original) for sine wave reconstruction example, and it comes out to be 1e-4 for the setting, but if I set M to 1, 2 or 8, while keeping taps=62, and using optimal cutoff ratio, I am getting imperfect reconstruction with high MSE error at 0.4789. I wonder why subbands = 4 is special.
